File: ios/GeelabGuardSdk.swift
import Foundation
import GeelabGuardSDK

struct GeelabGuardNativeReceipt {
  let appId: String?
  let geeToken: String?
  let geeId: String?
  let geeIdTimestamp: String?
  let respondedGeeToken: String?
  let originalResponse: Data?
}

protocol GeelabGuardSdk {
  func register(appId: String, serverUrl: String?)

  func fetchReceipt(signData: Data?) -> GeelabGuardNativeReceipt?

  func submitReceipt(
    signData: Data?,
    completion: @escaping (GeelabGuardNativeReceipt?, NSError?) -> Void
  )

  var version: String { get }
}

final class IOSGeelabGuardSdk: GeelabGuardSdk {
  func register(appId: String, serverUrl: String?) {
    if let serverUrl = serverUrl {
      GeelabGuardSDK.GeelabGuard.register(
        withAppID: appId,
        serverURL: serverUrl
      )
    } else {
      GeelabGuardSDK.GeelabGuard.register(withAppID: appId)
    }
  }

  func fetchReceipt(signData: Data?) -> GeelabGuardNativeReceipt? {
    GeelabGuardSDK.GeelabGuard.fetchReceipt(withSign: signData)?.toBridgeReceipt()
  }

  func submitReceipt(
    signData: Data?,
    completion: @escaping (GeelabGuardNativeReceipt?, NSError?) -> Void
  ) {
    GeelabGuardSDK.GeelabGuard.submitReceipt(withSign: signData) { receipt, error in
      completion(
        receipt?.toBridgeReceipt(),
        error.map { $0 as NSError }
      )
    }
  }

  var version: String {
    GeelabGuardSDK.GeelabGuard.sdkVersion()
  }
}

private extension GeelabGuardSDK.GeelabGuardReceipt {
  func toBridgeReceipt() -> GeelabGuardNativeReceipt {
    GeelabGuardNativeReceipt(
      appId: appID,
      geeToken: geeToken,
      geeId: geeID,
      geeIdTimestamp: geeIDTimestamp,
      respondedGeeToken: respondedGeeToken,
      originalResponse: originalResponse
    )
  }
}

File: android/src/main/java/tech/geelab/reactnative/geelabguard/GeelabGuardSdk.kt
package tech.geelab.reactnative.geelabguard

import android.content.Context
import tech.geelab.core.GeelabGuard as NativeGeelabGuard
import tech.geelab.core.GeelabGuardReceipt as NativeGeelabGuardReceipt

internal data class GeelabGuardNativeReceipt(
  val appId: String?,
  val geeToken: String?,
  val geeId: String?,
  val geeIdTimestamp: String?,
  val respondedGeeToken: String?,
  val originalResponse: String?,
)

internal interface GeelabGuardSdk {
  fun register(appId: String, serverUrl: String?)

  fun fetchReceipt(signData: String): GeelabGuardNativeReceipt?

  fun submitReceipt(
    signData: String,
    callback: (Int, GeelabGuardNativeReceipt?) -> Unit,
  )

  fun version(): String
}

internal class AndroidGeelabGuardSdk(
  private val applicationContext: Context,
) : GeelabGuardSdk {
  override fun register(appId: String, serverUrl: String?) {
    if (serverUrl == null) {
      NativeGeelabGuard.register(applicationContext, appId)
    } else {
      NativeGeelabGuard.register(applicationContext, appId, serverUrl)
    }
  }

  override fun fetchReceipt(signData: String): GeelabGuardNativeReceipt? =
    NativeGeelabGuard.fetchReceipt(applicationContext, signData)?.toBridgeReceipt()

  override fun submitReceipt(
    signData: String,
    callback: (Int, GeelabGuardNativeReceipt?) -> Unit,
  ) {
    NativeGeelabGuard.submitReceipt(
      applicationContext,
      signData,
      NativeGeelabGuard.CallbackHandler { status, receipt ->
        callback(status?.toInt() ?: UNKNOWN_STATUS, receipt?.toBridgeReceipt())
      },
    )
  }

  override fun version(): String = NativeGeelabGuard.getVersion()

  private fun NativeGeelabGuardReceipt.toBridgeReceipt() = GeelabGuardNativeReceipt(
    appId = appID,
    geeToken = geeToken,
    geeId = geeID,
    geeIdTimestamp = geeIDTimestamp,
    respondedGeeToken = respondedGeeToken,
    originalResponse = originalResponse,
  )

  private companion object {
    const val UNKNOWN_STATUS = Int.MIN_VALUE
  }
}
